### PR TITLE
Implement calls to 'super()'

### PR DESCRIPTION
**Pull request by dcodeIO**
Passes a simple test and implements quite a few related checks. Still possible that not every single edge case is covered, but having this now and fixing any bugs as we go appears reasonable, considering that class support is still in the works anyway.

### CODE REVIEW

**Comment by MaxGraey** [comment]
It seems this fixing also #375, #245, #199 and #187.

File: src/compiler.ts [hunk 1]
```diff
@@ -6658,6 +6683,37 @@ export class Compiler extends DiagnosticEmitter {
...
+    var constructorInstance: Function | null = null;
+    while (currentClassInstance) {
+      constructorInstance = currentClassInstance.constructorInstance;
+      if (constructorInstance) break; // TODO: check visibility
+      currentClassInstance = currentClassInstance.base;
+    }
+
+    // if a constructor is present, allocate the necessary memory for `this` and call it
+    var expr: ExpressionRef;
+    if (constructorInstance) {
```
Comment (MaxGraey) [inline src/compiler.ts]: Is it make sense also check constructor's body and if it empty don't call constructor?

Comment (dcodeIO) [reply src/compiler.ts]: Might be possible. Actual constructor code is generated code, though, that can contain other instructions like for initializing fields. In this case, though, that won't happen because a non-zero `this` value is provided. Aside from that I'm not too happy with constructor handling as-is, yet (conditional allocate and zero this for self-alloc gives me headaces but is necessary in order to return a custom instance from a ctor, hmm).

Comment (MaxGraey) [reply src/compiler.ts]: of course this case should call constructor anyway:
```ts
class Foo {
   a: i32
   b: i32 = this.a
}
```
and probably this:
```ts
class Foo {
   constructor(public a: i32) {}
}
```

**Comment by MaxGraey** [comment]
Hmm, I'm not sure about first statement call rule. It make sense only is `this` or `super` keyword using before `super` call. But should be valid for this case:

```ts
constructor(a: i32) {
   val arg = a + 1;
   super(arg);
}
```

**Comment by dcodeIO** [comment]
Yeah, there are the ALLOCATES and CONDITIONALLY_ALLOCATES flags in flows for that. Will probably need a CALLS_SUPER one as well then to do it this way.

**Comment by MaxGraey** [comment]
> Maybe we could use
Yeah

**Comment by dcodeIO** [comment]
Seems

```ts
constructor(a: i32) {
   var arg = a + 1;
   super(arg);
}
```

is rejected by TS, but now legalized in AS with the flow changes above. Still needs a check when doing `super.something` that super has already been called.

**Comment by MaxGraey** [comment]
Hmm, really? That's interesting...

**Comment by MaxGraey** [comment]
Yeah, but PR for fixing this in TS already exists: https://github.com/Microsoft/TypeScript/pull/29374

**Comment by dcodeIO** [comment]
Yeah, I remember that this limitation annoyed me in the past. Especially given the dynamic nature of JavaScript this seems like an arbitrary restriction (not directly related to AS, where such a restriction might actually make sense).

File: src/program.ts [hunk 2]
```diff
@@ -2931,6 +2931,21 @@ export class Class extends Element {
     }
   }
 
+  /** Gets the first applicable constructor. */
+  getfirstConstructor(includingThis: bool = true): Function | null {
```
Comment (MaxGraey) [inline src/program.ts]: Is it call somewhere?

**Comment by dcodeIO** [comment]
One thing that's still particularly off is that derived classes currently inherit their super class's fields, and thus initialize super fields in each derived constructor, which is odd. Instead, each constructor should initialize its own class's fields right after bubbling up the entire prototype chain, calling all the overloaded constructors sequentially. One reason why this doesn't yet work is that classes that don't have an explicit constructor also don't have a wasm function for it (their field initialization is instead inlined), which we should change by generating actual constructor functions for them that we can call.

**Comment by MaxGraey** [comment]
Makes sense. But it look like not simple. May be better leave this for other PR?

**Comment by dcodeIO** [comment]
Similarly:

```ts
class A {
  a: i32 = 1;
  constructor() {}
}

class B extends A {
  b: i32 = 2;
}

var b = new B();
assert(b.b == 2); // == 0, uninitialized due to A#constructor being used, with no B#constructor
```

File: src/compiler.ts (hunk 1 above)
Comment (dcodeIO) [reply src/compiler.ts]: Not going to optimize this yet as it would conflict with redoing constructors as mentioned below. But it's on TODO.

**Comment by dcodeIO** [comment]
Merging, continuing in a separate PR.

**Comment by MaxGraey** [comment]
@robrohan Basic OO supports except virtual methods. Issue which you mentioned is not yet fixed. If you add some params to constructor it would work
